File: go/find_minimum_in_rotated_sorted_array.go
package leetcode

// Time complexity: O(log n) where n is the length of nums
// Space complexity: O(1)
func findMin(nums []int) int {
	if nums[0] < nums[len(nums)-1] || len(nums) == 1 {
		return nums[0]
	}

	left, right, res := 0, len(nums)-1, nums[0]

	for left <= right {
		mid := left + (right-left)/2

		if nums[mid] < nums[0] { // right side
			res = min(res, nums[mid])
			right = mid - 1
		} else { // left side
			left = mid + 1
		}
	}
	return res
}

func min(x, y int) int {
	if x < y {
		return x
	} else {
		return y
	}
}

File: python/find-minimum-in-rotated-sorted-array.py
from typing import List

class Solution:
    # Time complexity: O(log n) where n is the length of nums
    # Space complexity: O(1)
    def findMin(self, nums: List[int]) -> int:
        if nums[0] < nums[-1] or len(nums) == 1: return nums[0]
        
        left, right = 0, len(nums)-1
        res = float("inf")
        
        while left <= right:
            mid = left + (right - left) // 2
            
            if nums[mid] <= nums[-1]: # right side
                res = min(res, nums[mid])
                right = mid-1
            else: # left side
                left = mid+1
        return res